File: 2024/python/day2.py
from collections import Counter
import re

def is_ordered(nums):
    s = sorted(nums)
    r = sorted(nums, reverse=True)
    return nums == s or nums == r

def is_safe(nums):
    if not is_ordered(nums):
        return False
    for i in range(1, len(nums)):
        d = abs(nums[i-1] - nums[i])
        if d == 0 or d > 3:
            return False
    return True

def try_all(nums, func):
    if func(nums):
        return True
    for i in range(len(nums)):
        if func(nums[:i] + nums[i+1:]):
            return True
    return False

part1 = 0
part2 = 0

with open(f"2024/_input/day2.txt") as fp:
    for line in fp.readlines():
        nums = [int(n) for n in line.split()]
        if is_safe(nums):
            part1 += 1
        if try_all(nums, is_safe):
            part2 += 1
        
print('part 1:', part1)
print('part 2:', part2)

File: 2024/go/day2.go
package main

import (
	"bufio"
	"fmt"
	"math"
	"os"
	"strconv"
	"strings"
)

func isOrdered(nums []int) bool {
	ascending := true
	descending := true

	for i := 1; i < len(nums); i++ {
		if nums[i-1] > nums[i] {
			ascending = false
		}
		if nums[i-1] < nums[i] {
			descending = false
		}
	}

	return ascending || descending
}

func isSafe(nums []int) bool {
	if !isOrdered(nums) {
		return false
	}

	for i := 1; i < len(nums); i++ {
		d := int(math.Abs(float64(nums[i-1] - nums[i])))
		if d == 0 || d > 3 {
			return false
		}
	}

	return true
}

func tryAll(nums []int, checkFunc func([]int) bool) bool {
	if checkFunc(nums) {
		return true
	}

	for i := range nums {
		newNums := make([]int, 0, len(nums)-1)
		newNums = append(newNums, nums[:i]...)
		newNums = append(newNums, nums[i+1:]...)

		if checkFunc(newNums) {
			return true
		}
	}

	return false
}

func readInput() [][]int {
	file, err := os.Open("2024/_input/day2.txt")
	if err != nil {
		fmt.Println("Error opening file:", err)
		os.Exit(1)
	}
	defer file.Close()

	scanner := bufio.NewScanner(file)

	result := make([][]int, 0)
	for scanner.Scan() {
		line := scanner.Text()
		// Split line into numbers
		strNums := strings.Fields(line)
		nums := make([]int, len(strNums))

		// Convert string numbers to integers
		for i, str := range strNums {
			num, err := strconv.Atoi(str)
			if err != nil {
				fmt.Println("Error converting string to number:", err)
				continue
			}
			nums[i] = num
		}
		result = append(result, nums)
	}

	if err := scanner.Err(); err != nil {
		fmt.Println("Error reading file:", err)
		os.Exit(1)
	}

	return result
}

func main() {
	var part1, part2 int
	for _, nums := range readInput() {
		if isSafe(nums) {
			part1++
		}
		if tryAll(nums, isSafe) {
			part2++
		}
	}
	fmt.Println("part 1:", part1)
	fmt.Println("part 2:", part2)
}
